sidecar/discovery/discovery.go

```diff
@@ -159,3 +198,247 @@ func filterInstances(instances []*api.ServiceInstance, tags []string) []*api.Ser
+					cluster.OutlierDetection.ConsecutiveError = backend.Resilience.ConsecutiveErrors
+				}
+				if backend.Resilience.DetectionInterval > 0 {
+					cluster.OutlierDetection.IntervalMS = int(backend.Resilience.DetectionInterval * 1000)
+				}
+			}
+		}
+
+		clusters = append(clusters, cluster)
+
+	}
+
+	sort.Sort(envoy.ClustersByName(clusters))
+
+	return clusters
+}
+
+func (d *Discovery) getRoutes(w rest.ResponseWriter, req *rest.Request) {
+
+	instances, err := d.discovery.ListInstances()
+	if err != nil {
+		logrus.WithError(err).Error("Unable to retrieve instances")
+		w.WriteHeader(http.StatusServiceUnavailable)
+		return
+	}
+
+	ruleSet, err := d.rules.ListRules(&api.RuleFilter{})
+	if err != nil {
+		logrus.WithError(err).Error("Unable to retrieve rules")
+		w.WriteHeader(http.StatusServiceUnavailable)
+		return
+	}
+
+	envoy.SanitizeRules(ruleSet.Rules)
+	rules := AddDefaultRouteRules(ruleSet.Rules, instances)
+
+	routes := buildRoutes(rules)
+
+	respJSON := VirtualHosts{
+		VirtualHosts: []envoy.VirtualHost{
+			{
+				Name:    "backend",
+				Domains: []string{"*"},
+				Routes:  routes,
+			},
+		},
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.WriteJson(&respJSON)
+}
+
+func buildRoutes(ruleList []api.Rule) []envoy.Route {
+	routes := []envoy.Route{}
+	for _, rule := range ruleList {
+		if rule.Route != nil {
+			var headers []envoy.Header
+			if rule.Match != nil {
+				headers = make([]envoy.Header, 0, len(rule.Match.Headers))
+				for k, v := range rule.Match.Headers {
+					headers = append(
+						headers,
+						envoy.Header{
+							Name:  k,
+							Value: v,
+							Regex: true,
+						},
+					)
+				}
+			}
+
+			for _, backend := range rule.Route.Backends {
+				var path, prefix, prefixRewrite string
+				if backend.URI != nil {
+					path = backend.URI.Path
+					prefix = backend.URI.Prefix
+					prefixRewrite = backend.URI.PrefixRewrite
+				} else {
+					prefix = fmt.Sprintf("/%v/", backend.Name)
+					prefixRewrite = "/"
+				}
+
+				clusterName := envoy.BuildServiceKey(backend.Name, backend.Tags)
+
+				runtime := &envoy.Runtime{
+					Key:     envoy.BuildWeightKey(backend.Name, backend.Tags),
+					Default: int(backend.Weight * 100),
+				}
+
+				route := envoy.Route{
+					Runtime:       runtime,
+					Path:          path,
+					Prefix:        prefix,
+					PrefixRewrite: prefixRewrite,
+					Cluster:       clusterName,
+					Headers:       headers,
+					RetryPolicy: envoy.RetryPolicy{
+						Policy: "5xx,connect-failure,refused-stream",
+					},
+				}
+
+				if rule.Route.HTTPReqTimeout > 0 {
+					// convert from float sec to in ms
+					route.TimeoutMS = int(rule.Route.HTTPReqTimeout * 1000)
+				}
+				if rule.Route.HTTPReqRetries > 0 {
+					route.RetryPolicy.NumRetries = rule.Route.HTTPReqRetries
+				}
+
+				routes = append(routes, route)
+			}
+		}
+	}
+
+	return routes
+}
+
+// AddDefaultRouteRules TODO same as the one in envoy manager but takes []*api.ServiceInstance instead of []api.ServiceInstance
```

Comment:
@rshriram: If it makes the logic easier, I agree. Otherwise that can go in its own PR.